---
key_info:
  - "Assembly Name": "GU_Smar_1.1.8"
  - "Assembly Type": "Haploid"
  - "Assembly Level": "Contig"
  - "Genome Representation": "Full"
  - "Accession": "GCA_030871285.1"


# The params below were auto-generated, you should not need to edit them...
# unless you were warned by the add-new-species.py script.
title: "Genome assembly"
layout: "species_assembly"
url: "skeletonema_marinoi/assembly"
weight: 2

stats_data_path: "skeletonema_marinoi/species_stats"
lineage_data_path: "skeletonema_marinoi/taxonomy"
---

|||||| Content divider - do not remove ||||||

Notes: Assembly statistics and BUSCO percentages were calculated for the primary genome assembly GCA_030871285.1. Annotation statistics were calculated using [EDIT].

BUSCO notation: C: Complete; S: Single-copy; D: Duplicated; F: Fragmented; M: Missing; n: Total BUSCO genes included in the dataset (here: [EDIT]). See also [the official BUSCO manual](https://busco.ezlab.org/busco_userguide.html#interpreting-the-results).

### Publication(s)

### Methods

#### Samples

#### Genome assembly

#### Genome annotation

### Funding

### Acknowledgements
