---
# The params below were auto-generated, you should not need to edit them...
# unless you were warned by the add-new-species.py script.
title: "*Blastobotrys malaysiensis* NRRL Y-6417"
subtitle: "An Ascomycota yeast species"
last_updated: "05/05/2025" # format DD/MM/YYYY

layout: "species_intro"
banner_title: "Species overview"
weight: 1

url: "blastobotrys_malaysiensis"
science_name: "Blastobotrys malaysiensis"
lineage_data_path: "blastobotrys_malaysiensis/taxonomy"

cover_image: "/img/species/blastobotrys_malaysiensis.webp"
img_attrib_text: "Blastobotrys malaysiensis NRRL Y-6417 from Ferreira de Sá et al. (2022), CC BY 4.0"
img_attrib_link: "https://doi.org/10.3897/BDJ.10.e80226"

gbif_taxon_id: "3480481"
goat_webpage: "https://goat.genomehubs.org/record?recordId=378047&result=taxon&taxonomy=ncbi#blastobotrys%20malaysiensis"

# Optional params below,
# comment them in if you want to include them.
# iucn_category: "[EDIT]"
# iucn_link: "[EDIT]"

# swe_red_list: "[EDIT]"
# swe_red_list_link: "[EDIT]"
---

### Description

_Blastobotrys malaysiensis_ belongs to a category of species that are referred to within yeast biotechnology as _non-conventional_ yeasts (Mukherjee et al., 2017). Here, _non-conventional_ implies that they might have other phenotypes compared to the industrial workhorse _Saccharomyces cerevisiae_, such as improved tolerance to process conditions, or different metabolic activities. Yeasts from the _Blastobotrys_ genus have been named to of interest for biotechnological applications due their oleaginous (lipid-accumulating) nature and their broad substrate range (Sanya et al., 2021). By genomic analysis of 332 Ascomycete genomes, _B. illinoisensis_ NRRL YB-1343 (also known as CBS 10337), B. malaysiensis_, and four other species were discovered to have xylanolytic activity (Ravn et al., 2021).

The _B. malaysiensis_ NRRL Y-6417 genome assembly that is displayed here on the Genome Portal was generated in the Y1000+ Project (Opulente et al. 2024, Shen et al., 2018). The annotation of protein-coding genes was generated in Sweden in a study by the Geijer research group at Chalmers University of Technology.

#### How to cite

If you use the genome assembly presented in the genome portal from this species in your research, please cite the original publication:

```{style=citation}
Opulente, D. A., LaBella, A. L., Harrison, M.-C., Wolters, J. F., Liu, C., Li, Y., Kominek, J., Steenwyk, J. L., Stoneman, H. R., VanDenAvond, J., Miller, C. R., Langdon, Q. K., Silva, M., Gonçalves, C., Ubbelohde, E. J., Li, Y., Buh, K. V., Jarzyna, M., Haase, M. A. B., … Hittinger, C. T. (2024). Genomic factors shape carbon and nitrogen metabolic niche breadth across Saccharomycotina yeasts. Science, 384(6694), eadj4503. https://doi.org/10.1126/science.adj4503
```

If you use the annotation of protein-coding genes, which was generated by a different research group, please cite:

```{style=citation}
Ravn, J.; Sörensen Ristinmaa, A.; Mazurkewich, ScotS.t; Borges Dias, G.; Larsbrink, J.; Geijer, C. (2025). Gene annotation of _Blastobotrys mokoenaii_, Blastobotrys illinoisensis, and Blastobotrys malaysiensis. Chalmers University of Technology. Dataset. https://doi.org/10.17044/scilifelab.28606814.v1
```

If you have used the pages for this species in the Genome Portal, please refer to it in-text as: "The _Blastobotrys malaysiensis_ entry in the Swedish Reference Genome Portal (Retrieved <span class="todays-date"></span>)." and use the following for the bibliography:

```{style=citation}
Swedish Reference Genome Portal (Retrieved DATE_ACCESSED), SciLifeLab Data Centre, version VERSION_NUMBER from <https://genomes.scilifelab.se>, [RRID:SCR_026008](https://rrid.site/data/record/nlx_144509-1/SCR_026008/resolver?q=rrid:scr_026008)
```

#### References

- Ferreira de Sá A.S., Leonardo-Silva L., Xavier-Santos S. (2022). Expanding the geographical distribution of _Blastobotrys malaysiensis_ (Saccharomycetales) beyond the Asian continent – a cave fungus first reported in the Americas. Biodiversity Data Journal 10: e80226. <https://doi.org/10.3897/BDJ.10.e80226>

- Opulente, D. A., LaBella, A. L., Harrison, M.-C., Wolters, J. F., Liu, C., Li, Y., Kominek, J., Steenwyk, J. L., Stoneman, H. R., VanDenAvond, J., Miller, C. R., Langdon, Q. K., Silva, M., Gonçalves, C., Ubbelohde, E. J., Li, Y., Buh, K. V., Jarzyna, M., Haase, M. A. B., … Hittinger, C. T. (2024). Genomic factors shape carbon and nitrogen metabolic niche breadth across Saccharomycotina yeasts. Science, 384(6694), eadj4503. <https://doi.org/10.1126/science.adj4503>

- - Ravn, J. L., Engqvist, M. K. M., Larsbrink, J., & Geijer, C. (2021). CAZyme prediction in ascomycetous yeast genomes guides discovery of novel xylanolytic species with diverse capacities for hemicellulose hydrolysis. Biotechnology for Biofuels, 14(1), 150. <https://doi.org/10.1186/s13068-021-01995-x>

- Ravn, J. L., Sörensen Ristinmaa, A., Coleman, T., Larsbrink, J., & Geijer, C. (2023). Yeasts Have Evolved Divergent Enzyme Strategies To Deconstruct and Metabolize Xylan. Microbiology Spectrum, 11(3), e00245-23. <https://doi.org/10.1128/spectrum.00245-23>

- Shen, X.-X., Opulente, D. A., Kominek, J., Zhou, X., Steenwyk, J. L., Buh, K. V., Haase, M. A. B., Wisecaver, J. H., Wang, M., Doering, D. T., Boudouris, J. T., Schneider, R. M., Langdon, Q. K., Ohkuma, M., Endoh, R., Takashima, M., Manabe, R., Čadež, N., Libkind, D., … Rokas, A. (2018). Tempo and Mode of Genome Evolution in the Budding Yeast Subphylum. Cell, 175(6), 1533-1545.e20. <https://doi.org/10.1016/j.cell.2018.10.023>

#### Changelog

- 05/05/2025 - Species first published on the Portal
