---
key_info:
  - "Assembly Name": "[EDIT]"
  - "Assembly Type": "[EDIT]"
  - "Assembly Level": "[EDIT]"
  - "Genome Representation": "[EDIT]"
  - "Accession": "[EDIT]"


# The params below were auto-generated, you should not need to edit them...
# unless you were warned by the add-new-species.py script.
title: "Genome assembly"
layout: "species_assembly"
url: "SPECIES_FOLDER/assembly"
weight: 2

stats_data_path: "SPECIES_FOLDER/species_stats"
lineage_data_path: "SPECIES_FOLDER/taxonomy"
---

|||||| Content divider - do not remove ||||||

Notes: Assembly statistics were calculated for the primary genome assembly [EDIT:accession_number]. Busco statistics were taken from [EDIT:publicaion_reference]. Annotation statistics were calculated using [EDIT:filename].

BUSCO notation: C: Complete; S: Single-copy; D: Duplicated; F: Fragmented; M: Missing; n: Total BUSCO genes included in the dataset (here: [EDIT:busco_dataset]). See also [the official BUSCO manual](https://busco.ezlab.org/busco_userguide.html#interpreting-the-results).

### Publication(s)

The data for *[EDIT:species_name]* displayed in the genome portal comes from:

- <p> [EDIT] <a href="[EDIT:DOI_URL]"> [EDIT:DOI_number]]</a></p>

### Funding

*The study in which the genome data was generated ([EDIT:publicaion_reference]) acknowledge funding by:*

- [EDIT]
- [EDIT]