---
key_info:
  - "Assembly Name": "[EDIT]"
  - "Assembly Type": "[EDIT]"
  - "Assembly Level": "[EDIT]"
  - "Genome Representation": "[EDIT]"
  - "Accession": "[EDIT]"


# The params below were auto-generated, you should not need to edit them...
# unless you were warned by the add-new-species.py script.
title: "Genome assembly"
layout: "species_assembly"
url: "SPECIES_FOLDER/assembly"
weight: 2

stats_data_path: "SPECIES_FOLDER/species_stats"
lineage_data_path: "SPECIES_FOLDER/taxonomy"
---

|||||| Content divider - do not remove ||||||

Notes: Assembly statistics were calculated for the primary genome assembly [EDIT ACCESSION NUMBER]. Busco statistics were taken from [EDIT ADD PUBLICATION]. Annotation statistics were calculated using [EDIT ADD FILE NAME].

BUSCO notation: C: Complete; S: Single-copy; D: Duplicated; F: Fragmented; M: Missing; n: Total BUSCO genes included in the dataset (here: [EDIT BUSCO DATABASE NAME]). See also [the official BUSCO manual](https://busco.ezlab.org/busco_userguide.html#interpreting-the-results).

### Publication(s)

The data for *[ADD SPECIES NAME]* displayed in the genome portal comes from:

- [EDIT ADD PUBLICATION]

### Methods

*Below is a brief summary of the methodology used to produce the genome data, based on [ADD INFORMATION ON PUBLICATION HERE].*

#### Samples

- [EDIT]

#### Genome assembly

- [EDIT]

#### Genome annotation

- [EDIT]

### Funding

- [EDIT]

### Acknowledgements

- [EDIT]
