---
# The params below were auto-generated, you should not need to edit them...
# unless you were warned by the add-new-species.py script.
key_info:
  - "Assembly Name": "ASSEMBLY_NAME"
  - "Assembly Type": "ASSEMBLY_TYPE"
  - "Assembly Level": "ASSEMBLY_LEVEL"
  - "Genome Representation": "GENOME_REPRESENTATION"
  - "Accession": "ASSEMBLY_ACCESSION"

title: "Genome assembly"
layout: "species_assembly"
url: "SPECIES_SLUG/assembly"
weight: 2

stats_data_path: "SPECIES_SLUG/species_stats"
lineage_data_path: "SPECIES_SLUG/taxonomy"
---

Notes: Assembly statistics were calculated for the primary genome assembly ASSEMBLY_ACCESSION. Busco statistics were taken from [EDIT]. Annotation statistics were calculated using [EDIT].

BUSCO notation: C: Complete; S: Single-copy; D: Duplicated; F: Fragmented; M: Missing; n: Total BUSCO genes included in the dataset (here: arthropoda_odb10). See also [the official BUSCO manual](https://busco.ezlab.org/busco_userguide.html#interpreting-the-results).

### Publication(s)

The data for *SPECIES_NAME* displayed in the genome portal comes from:

```{style=citation}
PUBLICATION
```

### Funding

*The study in which the genome data was generated ([EDIT:publicaion_reference]) acknowledge funding by:*

FUNDING

