---
# The params below were auto-generated, you should not need to edit them...
# unless you were warned by the add_new_species.py script.
key_info:
  - "Assembly Name": "ASM370576v3"
  - "Assembly Type": "haploid"
  - "Assembly Level": "contig"
  - "Genome Representation": "full"
  - "Accession": "GCA_003705765.3"

title: "Genome assembly"
layout: "species_assembly"
url: "blastobotrys_mokoenaii/assembly"
weight: 2

stats_data_path: "blastobotrys_mokoenaii/species_stats"
lineage_data_path: "blastobotrys_mokoenaii/taxonomy"
---

Notes: Assembly statistics were calculated for the primary genome assembly GCA_003705765.3 (GCA_003705765.3_ASM370576v3_genomic.fna.gz) using Quast (v5.2.0; Mikheenko et al. 2018). Busco statistics were calculated for GCA_003705765.3_ASM370576v3_genomic.fna.gz using Busco (v5.8.3; Manni et al., 2021). Annotation statistics were calculated for bmokoenaii_annotation.gff using AGAT (v1.4.1; Dainat, 2024).

BUSCO notation: C: Complete; S: Single-copy; D: Duplicated; F: Fragmented; M: Missing; n: Total BUSCO genes included in the dataset (here: arthropoda_odb10). See also [the official BUSCO manual](https://busco.ezlab.org/busco_userguide.html#interpreting-the-results).

### Publication(s)

The data for *Blastobotrys mokoenaii* NRRL Y-27120 displayed in the genome portal comes from:

- Genome assembly:

```{style=citation}
Shen, X.-X., Opulente, D. A., Kominek, J., Zhou, X., Steenwyk, J. L., Buh, K. V., Haase, M. A. B., Wisecaver, J. H., Wang, M., Doering, D. T., Boudouris, J. T., Schneider, R. M., Langdon, Q. K., Ohkuma, M., Endoh, R., Takashima, M., Manabe, R., Čadež, N., Libkind, D., … Rokas, A. (2018). Tempo and Mode of Genome Evolution in the Budding Yeast Subphylum. Cell, 175(6), 1533-1545.e20. https://doi.org/10.1016/j.cell.2018.10.023
```

- Annotation of protein-coding genes:

```{style=citation}
Ravn, J.; Sörensen Ristinmaa, A.; Mazurkewich, ScotS.t; Borges Dias, G.; Larsbrink, J.; Geijer, C. (2025). Gene annotation of _Blastobotrys mokoenaii_, Blastobotrys illinoisensis, and Blastobotrys malaysiensis. Chalmers University of Technology. Dataset. https://doi.org/10.17044/scilifelab.28606814.v1
```

The tools used by the Genome Portal staff to calculate the statistics shown on top of this page are described in:

- Dainat J. (2024). AGAT: Another Gff Analysis Toolkit to handle annotations in any GTF/GFF format.
(Version v1.4.1). Zenodo. <https://www.doi.org/10.5281/zenodo.3552717>

- Manni, M., Berkeley, M. R., Seppey, M., Simão, F. A., & Zdobnov, E. M. (2021). BUSCO Update: Novel and Streamlined Workflows along with Broader and Deeper Phylogenetic Coverage for Scoring of Eukaryotic, Prokaryotic, and Viral Genomes. Molecular Biology and Evolution, 38(10), 4647–4654. <https://doi.org/10.1093/molbev/msab199>

- Mikheenko, A., Prjibelski, A., Saveliev, V., Antipov, D., & Gurevich, A. (2018). Versatile genome assembly evaluation with QUAST-LG. Bioinformatics, 34(13), i142–i150. <https://doi.org/10.1093/bioinformatics/bty266>

### Funding

*The annotation of protein-coding genes (Ravn et al., 2025) acknowledge funding by:*

- Novo Nordisk Foundation (Grant number: [NNF22OC0079899](https://app.dimensions.ai/details/grant/grant.13909076))
