---
# The params below were auto-generated, you should not need to edit them...
# unless you were warned by the add_new_species.py script.
key_info:
  - "Assembly Name": Euphydryas_editha
  - "Assembly Type": haploid
  - "Assembly Level": scaffold
  - "Genome Representation": full
  - "Accession": GCA_933666425.1

title: "Genome assembly"
layout: "species_assembly"
url: "euphydryas_editha/assembly"
weight: 2

stats_data_path: "euphydryas_editha/species_stats"
lineage_data_path: "euphydryas_editha/taxonomy"
---

Notes: Assembly statistics were calculated for the primary genome assembly GCA_933666425.1 (CAKOGL01.fna.gz) using Quast (v5.2.0; Mikheenko et al. 2018). Busco statistics (Manni et al., 2021) were taken from Tunstrom et al. (2022). Annotation statistics were calculated using CA_933666425.1_Euphydryas_editha_genomic.gff.gz using AGAT (v1.4.1; Dainat, 2024).

BUSCO notation: C: Complete; S: Single-copy; D: Duplicated; F: Fragmented; M: Missing; n: Total BUSCO genes included in the dataset (here: arthropoda_odb10). See also [the official BUSCO manual](https://busco.ezlab.org/busco_userguide.html#interpreting-the-results).

### Publication(s)

The data for *Euphydryas editha* displayed in the genome portal comes from:

```{style=citation}
Tunstrom, K., Wheat, C. W., Parmesan, C., Singer, M. C., & Mikheyev, A.
S. (2022). A Genome for Edith's Checkerspot Butterfly: An Insect with
Complex Host-Adaptive Suites and Rapid Evolutionary Responses to
Environmental Changes. Genome Biology and Evolution, 14(8), evac113.
https://doi.org/10.1093/gbe/evac11

```

The tools used by the Genome Portal staff to calculate the statistics shown on top of this page are described in:

- Dainat J. (2024). AGAT: Another Gff Analysis Toolkit to handle annotations in any GTF/GFF format.
(Version v1.4.1). Zenodo. <https://www.doi.org/10.5281/zenodo.3552717>

- Manni, M., Berkeley, M. R., Seppey, M., Simão, F. A., & Zdobnov, E. M. (2021). BUSCO Update: Novel and Streamlined Workflows along with Broader and Deeper Phylogenetic Coverage for Scoring of Eukaryotic, Prokaryotic, and Viral Genomes. Molecular Biology and Evolution, 38(10), 4647–4654. <https://doi.org/10.1093/molbev/msab199>

- Mikheenko, A., Prjibelski, A., Saveliev, V., Antipov, D., & Gurevich, A. (2018). Versatile genome assembly evaluation with QUAST-LG. Bioinformatics, 34(13), i142–i150. <https://doi.org/10.1093/bioinformatics/bty266>

### Funding

*The study in which the genome data was generated (Tunstrom et al. (2022)) acknowledge funding by:*

- Swedish Research Council (Grant number: [2017-04386](https://www.vr.se/swecris.html#/project/2017-04386_VR))
- The French Make Our Planet Great Again award (Project CCISS, number ANR-17-MPGA-0007)
- The Laboratoires d'Excellences (LABEX) TULIP (ANR-10-LABX-41)
- The Science for Life Laboratory National Project on Biodiversity (2014/R2-77)
