---
key_info:
  - "Assembly Name": "Mnor_HA_v2"
  - "Assembly Type": "Haploid"
  - "Assembly Level": "Scaffold"
  - "Genome Representation": "Full"
  - "Accession": "GCA_964058975.1"


# The params below were auto-generated, you should not need to edit them...
# unless you were warned by the add-new-species.py script.
title: "Genome assembly"
layout: "species_assembly"
url: "meganyctiphanes_norvegica/assembly"
weight: 2

stats_data_path: "meganyctiphanes_norvegica/species_stats"
lineage_data_path: "meganyctiphanes_norvegica/taxonomy"
---

Notes: Assembly statistics were calculated for the primary genome assembly GCA_964058975.1 (CAXKWB01.fasta.gz) using Quast (v5.2.0; Mikheenko et al. 2018). Busco statistics (Manni et al., 2021) were taken from Unneberg et al. (2024). Annotation statistics were calculated for GCA_964058975.1_Mnor_HA_v2_genomic.gff using AGAT (v1.4.1; Dainat, 2024).

BUSCO notation: C: Complete; S: Single-copy; D: Duplicated; F: Fragmented; M: Missing; n: Total BUSCO genes included in the dataset (here: arthropoda_odb10). See also [the official BUSCO manual](https://busco.ezlab.org/busco_userguide.html#interpreting-the-results).

### Publication(s)

The *Meganyctiphanes norvegica* data displayed in the genome portal comes from:

```{style=citation}
Unneberg, P., Larsson, M., Olsson, A., Wallerman, O., Petri, A., Bunikis, I., Vinnere Pettersson, O., Papetti, C., Gislason, A., Glenner, H., Cartes, J. E., Blanco-Bercial, L., Eriksen, E., Meyer, B., & Wallberg, A. (2024). Ecological genomics in the Northern krill uncovers loci for local adaptation across ocean basins. Nature Communications, 15(1), 6297. <https://doi.org/10.1038/s41467-024-50239-7>
```

The tools used by to calculate the statistics shown on top of this page are described in:

- Dainat J. (2024). AGAT: Another Gff Analysis Toolkit to handle annotations in any GTF/GFF format.
(Version v1.4.1). Zenodo. <https://www.doi.org/10.5281/zenodo.3552717>

- Manni, M., Berkeley, M. R., Seppey, M., Simão, F. A., & Zdobnov, E. M. (2021). BUSCO Update: Novel and Streamlined Workflows along with Broader and Deeper Phylogenetic Coverage for Scoring of Eukaryotic, Prokaryotic, and Viral Genomes. Molecular Biology and Evolution, 38(10), 4647–4654. <https://doi.org/10.1093/molbev/msab199>

- Mikheenko, A., Prjibelski, A., Saveliev, V., Antipov, D., & Gurevich, A. (2018). Versatile genome assembly evaluation with QUAST-LG. Bioinformatics, 34(13), i142–i150. <https://doi.org/10.1093/bioinformatics/bty266>

### Funding

*The study in which the genome data was generated (Unneberg et al. 2024) was funded by:*

- Swedish Research Council Formas [2017-00413](https://www.vr.se/english/swecris.html#/project/2017-00413_Formas)

- U.S. National Science Foundation, Division Of Ocean Sciences (NSF OCE) grants [1316040](https://www.nsf.gov/awardsearch/showAward?AWD_ID=1316040) and [1948162](https://www.nsf.gov/awardsearch/showAward?AWD_ID=1948162)

- Knut and Alice Wallenberg Foundation as part of the National Bioinformatics Infrastructure Sweden at SciLifeLab, grant ID KAW 2017.0003.
