---
key_info:
  - "Assembly Name": "US_GU_Lsax_2.0"
  - "Assembly Type": "Haploid"
  - "Assembly Level": "Chromosome"
  - "Genome Representation": "Full"
  - "Accession": "GCA_037325665.1"


# The params below were auto-generated, you should not need to edit them...
# unless you were warned by the add-new-species.py script.
title: "Genome assembly"
layout: "species_assembly"
url: "littorina_saxatilis/assembly"
weight: 2

stats_data_path: "littorina_saxatilis/species_stats"
lineage_data_path: "littorina_saxatilis/taxonomy"
---

Notes: Assembly statistics were calculated for the primary genome assembly GCA_037325665.1 (GCA_037325665.1_US_GU_Lsax_2.0_genomic.fna.gz) using Quast (v5.2.0; Mikheenko et al. 2018). Busco statistics (Manni et al., 2021) were taken from Supplemental Table S1 in De Jode et al. (2024). Annotation statistics were calculated for GCA_037325665.1_US_GU_Lsax_2.0_genomic.gff.gz using AGAT (v1.4.1; Dainat, 2024).

BUSCO notation: C: Complete; S: Single-copy; D: Duplicated; F: Fragmented; M: Missing; n: Total BUSCO genes included in the dataset (here: mollusca_odb10). See also [the official BUSCO manual](https://busco.ezlab.org/busco_userguide.html#interpreting-the-results).

### Publication(s)

The data for *Littorina saxatilis* displayed in the genome portal comes from:

```{style=citation}
De Jode, A., Faria, R., Formenti, G., Sims, Y., Smith, T. P., Tracey, A., Wood, J. M. D., Zagrodzka, Z. B., Johannesson, K., Butlin, R. K., & Leder, E. H. (2024). Chromosome-scale Genome Assembly of the Rough Periwinkle *Littorina saxatilis*. Genome Biology and Evolution, 16(4), evae076. <https://doi.org/10.1093/gbe/evae076>
```

The tools used by to calculate the statistics shown on top of this page are described in:

- Dainat J. (2024). AGAT: Another Gff Analysis Toolkit to handle annotations in any GTF/GFF format.
(Version v1.4.1). Zenodo. <https://www.doi.org/10.5281/zenodo.3552717>

- Manni, M., Berkeley, M. R., Seppey, M., Simão, F. A., & Zdobnov, E. M. (2021). BUSCO Update: Novel and Streamlined Workflows along with Broader and Deeper Phylogenetic Coverage for Scoring of Eukaryotic, Prokaryotic, and Viral Genomes. Molecular Biology and Evolution, 38(10), 4647–4654. <https://doi.org/10.1093/molbev/msab199>

- Mikheenko, A., Prjibelski, A., Saveliev, V., Antipov, D., & Gurevich, A. (2018). Versatile genome assembly evaluation with QUAST-LG. Bioinformatics, 34(13), i142–i150. <https://doi.org/10.1093/bioinformatics/bty266>

### Funding

*The study in which the genome data was generated (De Jode et al. 2024) acknowledge funding by:*

- The European Research Council (grant agreement number [693030](https://cordis.europa.eu/project/id/693030)),
- The Swedish Research Council VR (grant agreement numbers [2018-03695](https://www.vr.se/swecris.html#/project/2018-03695_VR), [2020-05385](https://www.vr.se/swecris.html#/project/2020-05385_VR), and [2021-04191](https://www.vr.se/swecris.html#/project/2021-04191_VR))
- The European Union's Horizon 2020 research and innovation program under the Marie Sklodowska-Curie (grant agreement number [706376](https://cordis.europa.eu/project/id/706376))
- The Fundação para a Ciência e a Tecnologia (2020.00275.CEECIND and PTDC/BIA-EVL/1614/2021)
