---
key_info:
  - "Assembly Name": "Afil_fr2py"
  - "Assembly Type": "Haploid"
  - "Assembly Level": "Chromosome"
  - "Genome Representation": "Full"
  - "Accession": "GCA_039555335.1"


# The params below were auto-generated, you should not need to edit them...
# unless you were warned by the add-new-species.py script.
title: "Genome assembly"
layout: "species_assembly"
url: "amphiura_filiformis/assembly"
weight: 2

stats_data_path: "amphiura_filiformis/species_stats"
lineage_data_path: "amphiura_filiformis/taxonomy"
---

Notes: Assembly statistics were calculated for the primary genome assembly GCA_039555335.1 (JAZBNO01.fasta.gz) using Quast (v5.2.0; Mikheenko et al. 2018). Busco statistics (Manni et al., 2021) were taken from Parey et al. (2024) and communication with the authors. Annotation statistics were calculated for Afil_fr2py.wn.filt.final.chroms.sorted.gtf using AGAT (v1.4.1; Dainat, 2024).

BUSCO notation: C: Complete; S: Single-copy; D: Duplicated; F: Fragmented; M: Missing; n: Total BUSCO genes included in the dataset (here: arthropoda_odb10). See also [the official BUSCO manual](https://busco.ezlab.org/busco_userguide.html#interpreting-the-results).

### Publication(s)

The data for *Amphiura filiformis* displayed in the genome portal comes from:

```{style=citation}
Parey, E., Ortega-Martinez, O., Delroisse, J., Piovani, L., Czarkwiani, A., Dylus, D., Arya, S., Dupont, S., Thorndyke, M., Larsson, T., Johannesson, K., Buckley, K. M., Martinez, P., Oliveri, P., & Marlétaz, F. (2024). The brittle star genome illuminates the genetic basis of animal appendage regeneration. Nature Ecology & Evolution, 8(8), 1505–1521. <https://doi.org/10.1038/s41559-024-02456-y>
```

The tools used by to calculate the statistics shown on top of this page are described in:

- Dainat J. (2024). AGAT: Another Gff Analysis Toolkit to handle annotations in any GTF/GFF format.
(Version v1.4.1). Zenodo. <https://www.doi.org/10.5281/zenodo.3552717>

- Manni, M., Berkeley, M. R., Seppey, M., Simão, F. A., & Zdobnov, E. M. (2021). BUSCO Update: Novel and Streamlined Workflows along with Broader and Deeper Phylogenetic Coverage for Scoring of Eukaryotic, Prokaryotic, and Viral Genomes. Molecular Biology and Evolution, 38(10), 4647–4654. <https://doi.org/10.1093/molbev/msab199>

- Mikheenko, A., Prjibelski, A., Saveliev, V., Antipov, D., & Gurevich, A. (2018). Versatile genome assembly evaluation with QUAST-LG. Bioinformatics, 34(13), i142–i150. <https://doi.org/10.1093/bioinformatics/bty266>

### Funding

*The study in which the genome data was generated (Parey et al 2024) acknowledge funding by:*

- The [Centre for Marine Evolutionary Biology](http://www.cemeb.science.gu.se/) at the University of Gothenburg
- The [IMAGO (Infrastructure for MArine Genetic model Organisms)](https://www.gu.se/en/cemeb-marine-evolutionary-biology/marine-genetic-models) project
