---
# The params below were auto-generated, you should not need to edit them...
# unless you were warned by the add_new_species.py script.
key_info:
  - "Assembly Name": iyAntPlag1
  - "Assembly Type": haploid
  - "Assembly Level": contig
  - "Genome Representation": full
  - "Accession": GCA_963995485.1

title: "Genome assembly"
layout: "species_assembly"
url: "anthophora_plagiata/assembly"
weight: 2

stats_data_path: "anthophora_plagiata/species_stats"
lineage_data_path: "anthophora_plagiata/taxonomy"
---

Notes: Assembly statistics were calculated for the primary genome assembly GCA_963995485.1 (CAXEFL01.fasta) using Quast (v5.2.0; Mikheenko et al. 2018). Busco statistics (Manni et al., 2021) were recieved by communication with the authors. Annotation statistics were calculated for GCA_963995485.1_iyAntPlag1_genomic.gff using AGAT (v1.4.1; Dainat, 2024).

BUSCO notation: C: Complete; S: Single-copy; D: Duplicated; F: Fragmented; M: Missing; n: Total BUSCO genes included in the dataset (here: arthropoda_odb10). See also [the official BUSCO manual](https://busco.ezlab.org/busco_userguide.html#interpreting-the-results).

### Publication(s)

The data for *Anthophora plagiata* displayed in the genome portal comes from:

```{style=citation}
Taliadoros, D., Soares, A. E. R., Dias, G., Bunikis, I., Pippel, M., Olsson, A., Mosbech, M.-B., Heintz, J., Lager, N., Strand, A.-S., Pettersson, M., Pettersson, O. V., Lantz, H., Cederberg, B., Lindblad-Toh, K., & Webster, M. T. (2025). Genome Variation in Three *Anthophora* Bee Species Reflects Divergent Demographic Histories. Molecular Ecology, n/a(n/a), e70204. https://doi.org/10.1111/mec.70204
```

The tools used by to calculate the statistics shown on top of this page are described in:

- Dainat J. (2024). AGAT: Another Gff Analysis Toolkit to handle annotations in any GTF/GFF format.
(Version v1.4.1). Zenodo. <https://www.doi.org/10.5281/zenodo.3552717>

- Manni, M., Berkeley, M. R., Seppey, M., Simão, F. A., & Zdobnov, E. M. (2021). BUSCO Update: Novel and Streamlined Workflows along with Broader and Deeper Phylogenetic Coverage for Scoring of Eukaryotic, Prokaryotic, and Viral Genomes. Molecular Biology and Evolution, 38(10), 4647–4654. <https://doi.org/10.1093/molbev/msab199>

- Mikheenko, A., Prjibelski, A., Saveliev, V., Antipov, D., & Gurevich, A. (2018). Versatile genome assembly evaluation with QUAST-LG. Bioinformatics, 34(13), i142–i150. <https://doi.org/10.1093/bioinformatics/bty266>

### Funding

*The study in which the genome data was generated (Taliadoros et al 2025) acknowledge funding by:*

- Swedish Research Council; Grant number: [2020-06174_VR](https://www.vr.se/english/swecris.html#/project/2020-06174_VR)
- Science for Life Laboratory
- National Academic Infrastructure for Supercomputing in Sweden, SNIC 2022/5-333 and SNIC 2023/23-465.
