---
key_info:
  - "Assembly Name": "Ltrigynum_genome"
  - "Assembly Type": "Haploid"
  - "Assembly Level": "Chromosome"
  - "Genome Representation": "Full"
  - "Accession": "GCA_964030455.1"


# The params below were auto-generated, you should not need to edit them...
# unless you were warned by the add-new-species.py script.
title: "Genome assembly"
layout: "species_assembly"
url: "linum_trigynum/assembly"
weight: 2

stats_data_path: "linum_trigynum/species_stats"
lineage_data_path: "linum_trigynum/taxonomy"
---

Notes: Assembly statistics were calculated for the primary genome assembly GCA_964030455.1. (GCA_964030455.1_Ltrigynum_genome_genomic.fna.gz) using Quast (v5.2.0; Mikheenko et al. 2018). Busco statistics (Manni et al., 2021) were taken from Gutiérrez-Valencia et al. (2024). Annotation statistics were calculated for Ltrigynum_v1_genes.gtf using AGAT (v1.4.1; Dainat, 2024).

BUSCO notation: C: Complete; S: Single-copy; D: Duplicated; F: Fragmented; M: Missing; n: Total BUSCO genes included in the dataset (here: eukaryota_odb10). See also [the official BUSCO manual](https://busco.ezlab.org/busco_userguide.html#interpreting-the-results).

### Publication(s)

The data for *Linum trigynum* displayed in the genome portal comes from:

```{style=citation}
Gutiérrez-Valencia, J., Zervakis, P.-I., Postel, Z., Fracassetti, M., Losvik, A., Mehrabi, S., Bunikis, I., Soler, L., Hughes, P. W., Désamoré, A., Laenen, B., Abdelaziz, M., Pettersson, O. V., Arroyo, J., & Slotte, T. (2024). Genetic Causes and Genomic Consequences of Breakdown of Distyly in *Linum trigynum*. Molecular Biology and Evolution, 41(5), msae087. <https://doi.org/10.1093/molbev/msae087>
```

The tools used by to calculate the statistics shown on top of this page are described in:

- Dainat J. (2024). AGAT: Another Gff Analysis Toolkit to handle annotations in any GTF/GFF format.
(Version v1.4.1). Zenodo. <https://www.doi.org/10.5281/zenodo.3552717>

- Manni, M., Berkeley, M. R., Seppey, M., Simão, F. A., & Zdobnov, E. M. (2021). BUSCO Update: Novel and Streamlined Workflows along with Broader and Deeper Phylogenetic Coverage for Scoring of Eukaryotic, Prokaryotic, and Viral Genomes. Molecular Biology and Evolution, 38(10), 4647–4654. <https://doi.org/10.1093/molbev/msab199>

- Mikheenko, A., Prjibelski, A., Saveliev, V., Antipov, D., & Gurevich, A. (2018). Versatile genome assembly evaluation with QUAST-LG. Bioinformatics, 34(13), i142–i150. <https://doi.org/10.1093/bioinformatics/bty266>

### Funding

*The study in which the genome data was generated (Gutiérrez-Valencia et al. 2024) acknowledge funding by:*

- The European Research Council (ERC) under the European Union’s Horizon 2020 research and innovation programme (grant agreement number [757451](https://cordis.europa.eu/project/id/757451))
- The Swedish Research Council (grant numbers [2019-04452](https://www.vr.se/english/swecris.html#/project/2019-04452_VR) and [2023-04532](https://www.vr.se/english/swecris.html#/project/2023-04532_VR))
- The Erik Philip-Sörensen foundation
- The Bergströms foundation
- The Nilsson-Ehle foundation
- The Carl Trygger foundation
